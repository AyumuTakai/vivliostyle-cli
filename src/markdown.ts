import { StringifyMarkdownOptions, VFM } from '@vivliostyle/vfm';
import fs from 'fs';
import vfile, { VFile } from 'vfile';

export interface VSFile extends VFile {
  data: {
    title?: string;
    theme?: string;
  };
}

export type PreProcess = (filepath: string, contents: string) => string;

function getContents(
  filepath: string,
  preprocess: PreProcess[] | undefined,
): string {
  let contents = fs.readFileSync(filepath, 'utf8');
  if (contents && preprocess) {
    for (const proc of preprocess) {
      contents = proc(filepath, contents);
    }
  }
  return contents;
}

export function processMarkdown(
  filepath: string,
  options: StringifyMarkdownOptions = {},
  preprocess: PreProcess[] | undefined = undefined,
): VSFile {
  const vfm = VFM(options);
  const processed = vfm.processSync(
    vfile({ path: filepath, contents: getContents(filepath, preprocess) }),
  ) as VSFile;
  return processed;
}
