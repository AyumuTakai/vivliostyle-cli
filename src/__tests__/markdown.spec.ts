import path from 'upath';
import { PreProcess, processMarkdown, VSFile } from '../markdown';

it('test processMarkdown', () => {
  const filepath = path.resolve(__dirname, 'markdown.md');
  const processed = processMarkdown(filepath);
  expect(processed).toMatchSnapshot();
});

it('test processMarkdown with empty preprocess scripts array', () => {
  const filepath: string = path.resolve(__dirname, 'markdown.md');
  const preprocess: PreProcess[] = [];
  const processed: VSFile = processMarkdown(filepath, undefined, preprocess);
  expect(processed).toMatchSnapshot();
});

it('test processMarkdown with a preprocess script', () => {
  const filepath: string = path.resolve(__dirname, 'markdown.md');
  const preprocess: PreProcess[] = [
    (filepath: string, contents: string) => {
      contents = contents.replace(/preprocess/, 'preprocessed');
      return contents;
    },
  ];
  const processed: VSFile = processMarkdown(filepath, undefined, preprocess);
  expect(processed).toMatchSnapshot();
});

it('test processMarkdown with empty replace rule', () => {
  const filepath = path.resolve(__dirname, 'markdown.md');
  const options = {
    replace: [],
  };
  const processed = processMarkdown(filepath, options);
  expect(processed).toMatchSnapshot();
});

it('test processMarkdown with a replace rule', () => {
  const filepath = path.resolve(__dirname, 'markdown.md');
  const options = {
    replace: [
      {
        // 強制改ページ
        test: /replace/,
        match: ([], h: any) => {
          return h('span', 'replaced');
        },
      },
    ],
  };
  const processed = processMarkdown(filepath, options);
  expect(processed).toMatchSnapshot();
});
