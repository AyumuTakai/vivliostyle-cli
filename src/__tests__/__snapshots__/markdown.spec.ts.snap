// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test processMarkdown 1`] = `
VFile {
  "contents": "<!doctype html>
<html>
  <head>
    <meta charset=\\"utf-8\\">
    <title>markdown</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
  </head>
  <body>
    <section id=\\"headings\\">
      <h1>headings</h1>
      <p>preprocess</p>
      <p>replace</p>
    </section>
  </body>
</html>
",
  "cwd": "/Users/takaiayumu/IdeaProjects/vivliostyle-cli",
  "data": Object {
    "title": "headings",
    "toc": false,
  },
  "history": Array [
    "/Users/takaiayumu/IdeaProjects/vivliostyle-cli/src/__tests__/markdown.md",
  ],
  "messages": Array [],
}
`;

exports[`test processMarkdown with a preprocess script 1`] = `
VFile {
  "contents": "<!doctype html>
<html>
  <head>
    <meta charset=\\"utf-8\\">
    <title>markdown</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
  </head>
  <body>
    <section id=\\"headings\\">
      <h1>headings</h1>
      <p>preprocessed</p>
      <p>replace</p>
    </section>
  </body>
</html>
",
  "cwd": "/Users/takaiayumu/IdeaProjects/vivliostyle-cli",
  "data": Object {
    "title": "headings",
    "toc": false,
  },
  "history": Array [
    "/Users/takaiayumu/IdeaProjects/vivliostyle-cli/src/__tests__/markdown.md",
  ],
  "messages": Array [],
}
`;

exports[`test processMarkdown with a replace rule 1`] = `
VFile {
  "contents": "<!doctype html>
<html>
  <head>
    <meta charset=\\"utf-8\\">
    <title>markdown</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
  </head>
  <body>
    <section id=\\"headings\\">
      <h1>headings</h1>
      <p>preprocess</p>
      <p><span>replaced</span></p>
    </section>
  </body>
</html>
",
  "cwd": "/Users/takaiayumu/IdeaProjects/vivliostyle-cli",
  "data": Object {
    "title": "headings",
    "toc": false,
  },
  "history": Array [
    "/Users/takaiayumu/IdeaProjects/vivliostyle-cli/src/__tests__/markdown.md",
  ],
  "messages": Array [],
}
`;

exports[`test processMarkdown with empty preprocess scripts array 1`] = `
VFile {
  "contents": "<!doctype html>
<html>
  <head>
    <meta charset=\\"utf-8\\">
    <title>markdown</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
  </head>
  <body>
    <section id=\\"headings\\">
      <h1>headings</h1>
      <p>preprocess</p>
      <p>replace</p>
    </section>
  </body>
</html>
",
  "cwd": "/Users/takaiayumu/IdeaProjects/vivliostyle-cli",
  "data": Object {
    "title": "headings",
    "toc": false,
  },
  "history": Array [
    "/Users/takaiayumu/IdeaProjects/vivliostyle-cli/src/__tests__/markdown.md",
  ],
  "messages": Array [],
}
`;

exports[`test processMarkdown with empty replace rule 1`] = `
VFile {
  "contents": "<!doctype html>
<html>
  <head>
    <meta charset=\\"utf-8\\">
    <title>markdown</title>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
  </head>
  <body>
    <section id=\\"headings\\">
      <h1>headings</h1>
      <p>preprocess</p>
      <p>replace</p>
    </section>
  </body>
</html>
",
  "cwd": "/Users/takaiayumu/IdeaProjects/vivliostyle-cli",
  "data": Object {
    "title": "headings",
    "toc": false,
  },
  "history": Array [
    "/Users/takaiayumu/IdeaProjects/vivliostyle-cli/src/__tests__/markdown.md",
  ],
  "messages": Array [],
}
`;
